File: loadercomplex-fabric/ver17/src/main/java/com/enderzombi102/loadercomplex/fabric17/impl/item/FabricItem.java
package com.enderzombi102.loadercomplex.fabric17.impl.item;


import com.enderzombi102.loadercomplex.minecraft.item.Item;
import com.enderzombi102.loadercomplex.minecraft.util.Hand;
import com.enderzombi102.loadercomplex.fabric17.impl.block.FabricBlockstate;
import com.enderzombi102.loadercomplex.fabric17.impl.entity.FabricEntity;
import com.enderzombi102.loadercomplex.fabric17.impl.entity.FabricLivingEntity;
import com.enderzombi102.loadercomplex.fabric17.impl.entity.FabricPlayer;
import com.enderzombi102.loadercomplex.fabric17.impl.utils.BlockUtils;
import com.enderzombi102.loadercomplex.fabric17.impl.world.FabricWorld;
import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
import net.minecraft.block.BlockState;
import net.minecraft.entity.Entity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemGroup;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.util.ActionResult;
import net.minecraft.util.TypedActionResult;
import net.minecraft.util.UseAction;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.registry.Registry;
import net.minecraft.world.World;

public class FabricItem extends net.minecraft.item.Item {
	private final Item itemImpl;

	public FabricItem( Item item ) {
		super(
			new FabricItemSettings()
				.maxCount( item.maxCount )
				.maxDamage( item.maxDamage )
		);
		this.itemImpl = item;
		item.implementationItem = this;
	}

	// logic methods override

	@Override
	public void postProcessNbt(NbtCompound nbt ) {
		this.itemImpl.postProcesstag( nbt );
	}

	@Override
	public ActionResult useOnBlock( ItemUsageContext ctx ) {
		return ActionResult.valueOf(
			this.itemImpl.useOnBlock(
				new FabricWorld( ctx.getWorld() ),
				new FabricPlayer( ctx.getPlayer() ),
				BlockUtils.toPosition( ctx.getBlockPos() ),
				Hand.valueOf( ctx.getHand().name() ),
				com.enderzombi102.loadercomplex.minecraft.util.Direction.valueOf( ctx.getPlayerFacing().name() )
			).name()
		);
	}

	@Override
	public TypedActionResult<ItemStack> use(World world, PlayerEntity user, net.minecraft.util.Hand hand) {
		ItemStack stack = user.getStackInHand(hand);
		return new TypedActionResult<>(
			ActionResult.valueOf(
				this.itemImpl.use(
					new FabricWorld( world ),
					new FabricPlayer( user ),
					new FabricItemStack( stack )
				).name()
			),
			stack
		);
	}

	@Override
	public ItemStack finishUsing(ItemStack stack, World world, LivingEntity user) {
		return (
			(FabricItemStack) this.itemImpl.finishUsing(
				new FabricItemStack( stack ),
				new FabricWorld( world ),
				new FabricLivingEntity( user )
			)
		).getStack();
	}

	@Override
	public boolean postHit(ItemStack stack, LivingEntity target, LivingEntity attacker) {
		return this.itemImpl.postHit(
			new FabricItemStack( stack ),
			new FabricLivingEntity( target ),
			new FabricLivingEntity( attacker )
		);
	}

	@Override
	public boolean postMine(ItemStack stack, World world, BlockState state, BlockPos pos, LivingEntity miner) {
		return this.itemImpl.postMine(
			new FabricItemStack( stack ),
			new FabricWorld( world ),
			new FabricBlockstate( state ),
			BlockUtils.toPosition( pos ),
			new FabricLivingEntity( miner )
		);
	}

	@Override
	public boolean isSuitableFor(BlockState state) {
		return this.itemImpl.isEffectiveOn( new FabricBlockstate( state ) );
	}

	@Override
	public ActionResult useOnEntity(ItemStack stack, PlayerEntity user, LivingEntity entity, net.minecraft.util.Hand hand) {
		return this.itemImpl.useOnEntity(
			new FabricItemStack(stack),
			new FabricPlayer( user ),
			new FabricLivingEntity( entity ),
			Hand.valueOf( hand.name() )
		) ? ActionResult.SUCCESS : ActionResult.PASS;
	}

	@Override
	public void inventoryTick(ItemStack stack, World world, Entity entity, int slot, boolean selected) {
		this.itemImpl.inventoryTick(
			new FabricItemStack(stack),
			new FabricEntity( entity ),
			slot,
			selected
		);
	}

	@Override
	public void onCraft(ItemStack stack, World world, PlayerEntity player) {
		this.itemImpl.onCraft(
			new FabricItemStack(stack),
			new FabricPlayer( player )
		);
	}

	@Override
	public void onStoppedUsing(ItemStack stack, World world, LivingEntity user, int remainingUseTicks) {
		this.itemImpl.onStoppedUsing(
			new FabricItemStack(stack),
			new FabricWorld( world ),
			new FabricLivingEntity( user ),
			remainingUseTicks
		);
	}

	@Override
	protected boolean isIn(ItemGroup group) {
		ItemGroup itemGroup = this.getGroup();
		return itemGroup != null && ( group == ItemGroup.SEARCH || group == itemGroup );
	}

	// getter methods override

	// getMiningSpeedMultiplier

	@Override
	public float getMiningSpeedMultiplier(ItemStack stack, BlockState state) {
		return this.itemImpl.miningSpeedMultiplier;
	}

	@Override
	public boolean isDamageable() {
		return this.itemImpl.maxDamage > 0 && (!this.itemImpl.hasVariants || this.itemImpl.maxCount == 1);
	}

	@Override
	public UseAction getUseAction(ItemStack stack) {
		return UseAction.valueOf( this.itemImpl.useAction.name() );
	}

	@Override
	public int getMaxUseTime(ItemStack stack) {
		return this.itemImpl.maxUseTime;
	}

	@Override
	public int getEnchantability() {
		return this.itemImpl.enchantability;
	}

	@Override
	public boolean canRepair(ItemStack stack, ItemStack ingredient) {
		String material = this.itemImpl.repairMaterial.toString();
		if ( material != null ) {
			return Registry.ITEM
					.getId( ingredient.getItem() )
					.toString()
					.equals( material );
		}
		return false;
	}

	public Item getItemImpl() {
		return itemImpl;
	}
}

File: loadercomplex-fabric/src/ver171/kotlin/com/enderzombi102/loadercomplex/fabric171/impl/item/FabricItem.kt
package com.enderzombi102.loadercomplex.fabric171.impl.item

import com.enderzombi102.loadercomplex.api.minecraft.item.Item
import com.enderzombi102.loadercomplex.fabric171.impl.block.FabricBlockState
import com.enderzombi102.loadercomplex.fabric171.impl.entity.FabricEntity
import com.enderzombi102.loadercomplex.fabric171.impl.entity.FabricLivingEntity
import com.enderzombi102.loadercomplex.fabric171.impl.entity.FabricPlayerEntity
import com.enderzombi102.loadercomplex.fabric171.impl.utils.toLC
import com.enderzombi102.loadercomplex.fabric171.impl.utils.toMC
import com.enderzombi102.loadercomplex.fabric171.impl.world.FabricWorld
import net.fabricmc.fabric.api.item.v1.FabricItemSettings
import net.minecraft.block.BlockState
import net.minecraft.entity.Entity
import net.minecraft.entity.LivingEntity
import net.minecraft.item.ItemStack
import net.minecraft.item.ItemUsageContext
import net.minecraft.nbt.NbtCompound
import net.minecraft.util.ActionResult
import net.minecraft.util.TypedActionResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.registry.Registry
import net.minecraft.world.World
import net.minecraft.block.BlockState as McBlockState
import net.minecraft.entity.LivingEntity as McLivingEntity
import net.minecraft.entity.player.PlayerEntity as McPlayerEntity
import net.minecraft.item.Item as McItem
import net.minecraft.item.ItemStack as McItemStack
import net.minecraft.util.ActionResult as McActionResult
import net.minecraft.util.Hand as McHand
import net.minecraft.util.UseAction as McUseAction
import net.minecraft.world.World as McWorld


class FabricItem( private val itemImpl: Item ) : McItem( createSettings( itemImpl ) ) {
	init {
		itemImpl.implementationItem = this
//				if (this.itemImpl.maxDamage > 0) {
//			this.addModelPredicateProvider(
//					new Identifier("damaged"),
//					ItemAccessor.getDamagedProvider()
//			);
//			this.addModelPredicateProvider(
//					new Identifier("damage"),
//					ItemAccessor.getDamageProvider()
//			);
//		}
	}


	// region logic methods override
	override fun postProcessNbt( nbt: NbtCompound ) {
		itemImpl.validateTag( nbt )
	}

	override fun useOnBlock( ctx: ItemUsageContext ): McActionResult =
		itemImpl.useOnBlock(
			FabricWorld( ctx.world ),
			FabricPlayerEntity( ctx.player as McPlayerEntity ),
			ctx.blockPos.toLC(),
			ctx.hand.toLC(),
			ctx.side.toLC()
		).toMC()

	override fun use( world: McWorld, player: McPlayerEntity, hand: McHand ): TypedActionResult<McItemStack> {
		val holder = itemImpl.startUsing(
			FabricWorld( world ),
			FabricPlayerEntity( player ),
			FabricItemStack( player.getStackInHand( hand ) )
		)
		return TypedActionResult( holder.result.toMC(), holder.`object`.toMC() )
	}

	override fun finishUsing( stack: McItemStack, world: McWorld, entity: McLivingEntity ): McItemStack =
		itemImpl.finishUsing( FabricItemStack( stack ), FabricWorld( world ), FabricLivingEntity( entity ) ).toMC()

	override fun postHit( stack: McItemStack, target: McLivingEntity, attacker: McLivingEntity ): Boolean =
		itemImpl.postAttack( FabricItemStack( stack ), FabricLivingEntity( target ), FabricLivingEntity( attacker ) )

	override fun postMine( stack: McItemStack, world: McWorld, state: McBlockState, pos: BlockPos, entity: McLivingEntity): Boolean =
		itemImpl.postMine(
			FabricItemStack( stack ),
			FabricWorld(world),
			FabricBlockState( state ),
			pos.toLC(),
			FabricLivingEntity( entity )
		)

	override fun isSuitableFor( state: McBlockState ): Boolean =
		itemImpl.canMineBlock( FabricBlockState( state ) )

	override fun useOnEntity( stack: McItemStack, player: McPlayerEntity, entity: McLivingEntity, hand: McHand ): ActionResult =
		itemImpl.useOnEntity(
			FabricItemStack( stack ),
			FabricPlayerEntity( player ),
			FabricLivingEntity( entity ),
			hand.toLC()
		).toMC()

	override fun inventoryTick( stack: ItemStack, world: World, entity: Entity, slot: Int, selected: Boolean ) {
		itemImpl.inventoryTick(
			FabricItemStack( stack ),
			FabricEntity( entity ),
			slot,
			selected
		)
	}

	override fun onCraft( stack: McItemStack, world: McWorld, player: McPlayerEntity ) {
		itemImpl.onCraft( FabricItemStack( stack ), FabricPlayerEntity( player ) )
	}

	override fun onStoppedUsing( stack: ItemStack, world: World, user: LivingEntity, remainingUseTicks: Int ) {
		itemImpl.onStoppedUsing(
			FabricItemStack( stack ),
			FabricWorld( world ),
			FabricLivingEntity( user ),
			remainingUseTicks
		)
	}
	// endregion

	// region getter methods override
	override fun getMiningSpeedMultiplier(stack: ItemStack?, state: BlockState?): Float =
		itemImpl.miningSpeedMultiplier

	override fun isDamageable(): Boolean =
		itemImpl.maxDamage > 0 && (!itemImpl.hasVariants || itemImpl.maxCount == 1)

	override fun getUseAction( stack: McItemStack): McUseAction =
		itemImpl.useAction.toMC()

	override fun getEnchantability(): Int =
		itemImpl.enchantability

	override fun canRepair( stack: McItemStack, ingredient: McItemStack ): Boolean {
		val material = itemImpl.repairMaterial?.toString()

		return material != null && Registry.ITEM.getKey( ingredient.item ).toString() == material
	}
	// endregion

	companion object {
		fun createSettings( item: Item ): FabricItemSettings {
			val settings = FabricItemSettings()
				.maxCount( item.maxCount )
				.maxDamage( item.maxDamage )

			return settings
		}
	}
}
